### PR TITLE
[WIP][Not for landing now] TP benchmark for perf

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106409
* #107181
* #106524

### CODE REVIEW

File: torch/distributed/nn/functional.py [hunk 1]
```diff
@@ -101,6 +101,25 @@ def reduce_scatter(output, input_list, op=ReduceOp.SUM, group=group.WORLD):
     return _Reduce_Scatter.apply(op, group, output, *input_list)
 
 
+def _reduce_scatter_base(output, input, op=ReduceOp.SUM, group=group.WORLD):
```
Comment (wanchaol) [inline torch/distributed/nn/functional.py:104]: I think we should try to call the new functional collective instead of using the old functional collective?

Comment (fduwjj) [reply torch/distributed/nn/functional.py:104]: This is for control group which is to mimic Megatron.

File: benchmarks/distributed/tensor/tp_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls,
+            local_tensor.size(),
+            strides=local_tensor.stride(),
+            dtype=local_tensor.dtype,
+            device=local_tensor.device,
+            layout=local_tensor.layout,
+            requires_grad=local_tensor.requires_grad,
+        )
+        r._local_tensor = local_tensor.detach()
```
Comment (wanchaol) [inline benchmarks/distributed/tensor/tp_benchmark.py:71]: no need to detach it

File: benchmarks/distributed/tensor/tp_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        return r
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap_schema(e: object) -> object:
+            return e._local_tensor if isinstance(e, DummyDTensor) else e
+        args_unwrap = tree_map(unwrap_schema, args)
+        kwargs_unwrap = tree_map(unwrap_schema, kwargs)
+        return func(*args_unwrap, **kwargs_unwrap)
```
Comment (wanchaol) [inline benchmarks/distributed/tensor/tp_benchmark.py:80]: not only you need to unwrap it, you also need to rewrap it by calling another tree_map to convert the output to dummyDTensor

File: benchmarks/distributed/tensor/tp_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+def run_tp(rank, args):
+    """
+    Main body of the demo of a basic version of tensor parallel by using
+    PyTorch native APIs.
+    """
+    print(f"Running basic Megatron style TP example on rank {rank}.")
+    setup(rank, args.world_size)
+    tp_degree = args.world_size
+    dim_size = 64
```
Comment (wanchaol) [inline benchmarks/distributed/tensor/tp_benchmark.py:128]: I am wondering if we should mimic the real world run on dim size to conduct this experiment? i.e. we should figure out the 65B model MLP matmul shape and use that to measure, because when going large the compute time will also become larger, which could possibly affect the benchmarks.

Comment (fduwjj) [reply benchmarks/distributed/tensor/tp_benchmark.py:128]: yes, I can change to the size of MLP within 65B model.

File: benchmarks/distributed/tensor/tp_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    # Control group to mimic Megatron.
+    model = MLPModel(dim_size // tp_degree).cuda(rank)
+    optimizer = torch.optim.SGD(model.parameters(), lr=LR)
+    inp = torch.rand(128, 128).cuda(rank)
+    torch.distributed.barrier()
+    with maybe_run_profiler(args, "./control/") as torch_profiler:
+        for i in range(args.iter_nums):
+            if args.sequence_parallel:
+                output = torch.empty(inp.size(0) * tp_degree, *inp.size()[1:], device=inp.device)
+                input = _all_gather_base(output, inp)
```
Comment (davidberard98) [inline benchmarks/distributed/tensor/tp_benchmark.py:145]: do you need `else: input = inp` to make this run with `--sequence-parallel False`
